### PR TITLE
Issue #5264 Demo modules

### LINKED ISSUE #5264: Create demo module
**Jetty version**
10.0.x

**Description**
In light of the changes for Issue #5256, we need to migrate the existing demo-base to a new `demo` module.

This should probably exist in the `/examples/demo-module/` directory in source control.

It might make sense to also have the individual test webapps be modules themselves that the new `demo` module just depends on.
It could make adding a test webapp for request dump only purposes quite handy.

We should review the maven coordinates of all of the test webapps, as some of them are not aligned properly.

### PR DESCRIPTION
**Pull request by gregw**
Create modules in jetty-home to download and deploy demos for #5264

### CODE REVIEW

**Comment by gregw** [comment]
@sbordet Do you like this approach?

**Comment by joakime** [comment]
This could be done easier without including everything and the kitchen sink in jetty-home itself.
Why can't we use / unpack the `config` classified artifacts here? (why are we including the various xml that belongs to the various demos in jetty-home itself?)

**Comment by gregw** [comment]
@joakime this is a draft, just to see if we like how the approach works.
If the basic approach is OK, then yes assembling it from configs would be better. For now I didn't want to  break those configs and thus break distro, as I am using it as a comparison.

**Comment by gregw** [comment]
@joakime I'm in and out mostly for the next few days... so as this is on critical path for the release, can I pass it over to you?
@sbordet agrees with the basic approach as it lines up with his doco efforts well.  Thus what is left to do includes:
 + build a few more modules to make --add-module=demo-base produce more or less identical results to what we already have in demo-base.  I know I'm missing jaas and jndi webapps... but probably a few more things.
 + assemble those modules into jetty-home using the config mechanism - probably moving the mod files etc. back out to the jetty-xxx modules...
 + make sure that all the artifacts we get from maven will be published with a release.
 + either get rid of jetty-distribution... or better yet just make it use a `--add-module=demo-base` to create itself for now.

This will need lots of testing... so perhaps that's how I can help by tuning in and doing some testing and generally being a PITA.
If you can't take this hot potato, can you pass onto somebody else?

**Comment by joakime** [comment]
I've done a few things in the [jetty-10.0.x-5264-demo-with-start-extract](https://github.com/eclipse/jetty.project/compare/jetty-10.0.x-5264-demo-with-start-extract) branch.

Since I needed to test download from a repository, i added the following features to `jetty-start`.

* Now supports SNAPSHOT repositories properly (reads the `maven-metadata.xml` to figure out the snapshot version)
* Now supports extract on maven `[files]` section ... `maven://<groupId>/<artifactId>/<version>/<type>/<classfier>|extract:<path-in-jetty-base>`

This way we can use the `maven-deploy-plugin` to test a repository on the local disk (to know if we have everything properly deploying).

Example:
```
$ mkdir /tmp/jetty/snapshot-repo
$ cd /tmp/jetty/snapshot-repo
$ SNAPSHOT_REPO=`pwd`
$ cd /path/to/jetty.project-10.0.x
$ mvn clean install deploy -DskipTests -DaltSnapshotDeploymentRepository=jettysnapshots::default::file://$SNAPSHOTS_REPO
# and then later when testing ...
$ java -Dmaven.repo.uri=file://tmp/jetty/snapshot-repo -jar start.jar --add-to-start=demo-base
```

That branch also has a `home-config` assembly that will create `home` classified artifacts if the `src/main/home` directory exists.
This is so that the artifacts can overlay `jetty-home` specific configuration for the building of `jetty-home`.
That way we can keep the modules (and other files) in near their artifacts.

That branch also has a `/tests/test-webapps/demo-base-webapp` which is essentially the `webapps/ROOT` (courtesy of "-config.jar") and the `demo-base.mod` (courtesy of `-home.jar`)

I've started moving the demos files to their specific artifacts, but I'm not done.
I'll continue my efforts on monday.

**Comment by gregw** [comment]
@joakime why do we need to create a new home config concept?   All configs should now go to jetty-home and none should go to jetty-distribution.  So we should be able to continue using src/main/config and don't need a src/main/home (even though that is a better name :)

**Comment by joakime** [comment]
`src/main/config` is for the live jetty.base. (it's for any optional components, that don't ship with jetty-home, that are downloaded as part of a module)
`src/main/home` is for the jetty.home. (is for managing the jetty-home files, keeping the module, and it's relevant configs, near the artifact that being managed by source control)

**Comment by gregw** [comment]
@joakime I don't get the distinction you are trying to make?    `config` is for files that get expanded into `jetty-home` during the build process, which will now include the demo modules.
I just don't think we want or need to introduce an extra mechanism at this point.    We know more or less how we want jetty-home to look - this PR shows that.  We just need to use the existing config mechanism to assemble it.

**Comment by gregw** [comment]
@joakime so are you saying that instead of having some files in the `module/demo.d` file, you want to re-purpose the `config` mechanism to be published artefacts of config files that can be downloaded by modules and the existing `config` mechanism is being renamed to `home` and it is just used to assemble during a build?

Hmmmm I don't dislike it, but it just feels a bit too much to add at this very VERY late state, just to avoid a `module/demo.d` directory with a few xml and similar small files.  I'd much rather get this done with the features we have (although expand is a good addition).

**Comment by janbartel** [comment]
IIUC @joakime this is a proposal to allow demo modules to be downloaded? I'm in favour of having them be downloaded. I'm not in favour of a system whereby I download a jetty distribution and there's nothing in it essentially other than a means to download everything: I want to be able to look at the modules and the xml files and investigate them before I enable any of them.

I'm also very concerned about this new circular dependency that has arisen: it should absolutely be possible to cd to tests/test-webapps/test-jetty-webapp and just type mvn jetty:run. If we can't do that then something is wrong.

**Comment by gregw** [comment]
@joakime I think I understand the problem you are trying to solve.... it is a bit funny having the distributed jetty-home contain `module/demo-foo.mod` files and `modules/demo.d/demo-foo.xml` files, but then to download the war or libs used by the module.

But I don't think your solution helps as it breaks an optional module into 3 things:  1) the mod file baked into jetty-home; 2) the published artefact (war or lib); 3) the published config artefact

I better mid term solution would be to support remote modules.  Ie `demo-foo.mod` would not be part of jetty-home at all, nor any xml files.  Instead we could support `--add-module=http://some/url/to/the/demo-foo.module`, and this URL (or mvn coords) would download the `module/demo-foo.mod` files and `modules/demo.d/demo-foo.xml` files, maybe even the war and libs. This module could then be enabled, which might then download more files.

I think remote modules is a better solution to the problem you are trying to solve.... but too late for that feature.

Let's just go with a `modules/demo.d` directory containing a few xmls etc that is assembled with our current config mechanism.

File: jetty-start/src/test/java/org/eclipse/jetty/start/fileinits/MavenLocalRepoFileInitializerTest.java [hunk 1]
```diff
@@ -172,28 +177,90 @@ public void testGetCoordinateTestMavenBaseUri()
...
+
+    @Test
+    public void testDownloadSnapshotRepo()
+        throws Exception
+    {
+        Path snapshotLocalRepoDir = testdir.getPath().resolve("snapshot-repo");
+        FS.ensureEmpty(snapshotLocalRepoDir);
+
+        MavenLocalRepoFileInitializer repo =
+            new MavenLocalRepoFileInitializer(baseHome, snapshotLocalRepoDir, false, "https://oss.sonatype.org/content/repositories/jetty-snapshots/");
```
Comment (olamy) [inline jetty-start/src/test/java/org/eclipse/jetty/start/fileinits/MavenLocalRepoFileInitializerTest.java:213]: I don't like much the idea about having test depending on downloading resources from https://oss.sonatype.org.
@joakime not urgent but I can fix that (next week when back from holidays)

File: jetty-start/src/test/java/org/eclipse/jetty/start/fileinits/MavenMetadataTest.java [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+import org.xml.sax.SAXException;
+
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.is;
+import static org.junit.jupiter.api.Assertions.assertFalse;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+public class MavenMetadataTest
+{
+    private static final TimeZone UTC = TimeZone.getTimeZone("UTC");
```
Comment (olamy) [inline jetty-start/src/test/java/org/eclipse/jetty/start/fileinits/MavenMetadataTest.java]: variable not used

File: jetty-start/src/test/java/org/eclipse/jetty/start/fileinits/MavenMetadataTest.java [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+    @Test
+    public void testIsExpiredTimestampYesterday()
+    {
+        LocalDateTime yesterday = LocalDateTime.now().minusDays(1);
+        String timestamp = getTimestampFormatter().format(yesterday);
+        assertTrue(MavenMetadata.isExpiredTimestamp(timestamp), "Timestamp should be stale: " + timestamp);
+    }
+
+    @Test
+    public void testIsExpiredTimestamp12HoursOld()
```
Comment (olamy) [inline jetty-start/src/test/java/org/eclipse/jetty/start/fileinits/MavenMetadataTest.java]: this test will fail is executed before noon :)

File: jetty-start/src/test/java/org/eclipse/jetty/start/fileinits/MavenMetadataTest.java [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+    {
+        LocalDateTime oldish = LocalDateTime.now().minusHours(12);
```
Comment (olamy) [inline jetty-start/src/test/java/org/eclipse/jetty/start/fileinits/MavenMetadataTest.java]: what about something such `now(ZoneId.of("UTC"))` as `getTimestampFormatter` use `UTC`?

File: jetty-start/src/main/java/org/eclipse/jetty/start/fileinits/MavenLocalRepoFileInitializer.java [hunk 5]
```diff
@@ -219,6 +290,65 @@ public Coordinates getCoordinates(URI uri)
...
+        if (!Files.exists(localRepoMetadataPath))
+        {
+            // doesn't exist, it's stale.
+            return true;
+        }
+
+        try
+        {
+            MavenMetadata mavenMetadata = new MavenMetadata(localRepoMetadataPath);
+            return MavenMetadata.isExpiredTimestamp(mavenMetadata.getLastUpdated());
```
Comment (olamy) [inline jetty-start/src/main/java/org/eclipse/jetty/start/fileinits/MavenLocalRepoFileInitializer.java:330]: maybe we could have some configuration à la `mvn -U` to force downloading SNAPSHOT as here we have a sort of hardcoded value (the day before) maybe a sys property?

File: jetty-start/src/test/java/org/eclipse/jetty/start/fileinits/MavenLocalRepoFileInitializerTest.java (hunk 1 above)
Comment (joakime) [reply jetty-start/src/test/java/org/eclipse/jetty/start/fileinits/MavenLocalRepoFileInitializerTest.java:213]: We could setup a small remote repo in `src/test/resources` (a 1 artifact repo) for testing purposes.

File: jetty-start/src/test/java/org/eclipse/jetty/start/fileinits/MavenMetadataTest.java (hunk 3 above)
Comment (joakime) [reply jetty-start/src/test/java/org/eclipse/jetty/start/fileinits/MavenMetadataTest.java]: Yeah, i caught that 1 failure. :-)
I'm thinking of just dropping that specific test method.

File: jetty-start/src/test/java/org/eclipse/jetty/start/fileinits/MavenMetadataTest.java (hunk 4 above)
Comment (joakime) [reply jetty-start/src/test/java/org/eclipse/jetty/start/fileinits/MavenMetadataTest.java]: That might work.

File: jetty-start/src/main/java/org/eclipse/jetty/start/fileinits/MavenLocalRepoFileInitializer.java (hunk 5 above)
Comment (joakime) [reply jetty-start/src/main/java/org/eclipse/jetty/start/fileinits/MavenLocalRepoFileInitializer.java:330]: Forcing download of new snapshots could be useful, yes.
Lets make that a new PR after this one gets merged.

File: jetty-start/src/test/java/org/eclipse/jetty/start/fileinits/MavenMetadataTest.java (hunk 4 above)
Comment (joakime) [reply jetty-start/src/test/java/org/eclipse/jetty/start/fileinits/MavenMetadataTest.java]: I opted to just remove this specific test, it's not reliable as written, and is too sensitive to the time of day that surefire user executes the test on.

File: jetty-start/src/test/java/org/eclipse/jetty/start/fileinits/MavenLocalRepoFileInitializerTest.java (hunk 1 above)
Comment (joakime) [reply jetty-start/src/test/java/org/eclipse/jetty/start/fileinits/MavenLocalRepoFileInitializerTest.java:213]: I'll work on this today.

File: jetty-start/src/main/java/org/eclipse/jetty/start/fileinits/MavenLocalRepoFileInitializer.java (hunk 5 above)
Comment (joakime) [reply jetty-start/src/main/java/org/eclipse/jetty/start/fileinits/MavenLocalRepoFileInitializer.java:330]: This ability, to support SNAPSHOT artifacts, is currently only useful to us when testing things.
The actual 10.0.0 release wouldn't have this requirement.
So I think it's safe to push this `--update-snapshots` feature to a future Jetty 10.0.x release.

File: examples/async-rest/async-rest-webapp/src/main/config/modules/demo-async-rest.mod [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+
+[tags]
+demo
+webapp
+
+[depends]
+deploy
+
+[files]
+maven://org.eclipse.jetty.example-async-rest/example-async-rest-webapp/${jetty.version}/war|webapps/demo-async-rest.war
```
Comment (sbordet) [inline examples/async-rest/async-rest-webapp/src/main/config/modules/demo-async-rest.mod]: The `groupId` should be `org.eclipse.jetty.examples` like we have `o.e.j.tests`.

File: jetty-home/pom.xml [hunk 7]
```diff
@@ -464,7 +464,7 @@
               <goal>unpack-dependencies</goal>
             </goals>
             <configuration>
-              <includeGroupIds>org.eclipse.jetty,org.eclipse.jetty.websocket</includeGroupIds>
+              <includeGroupIds>org.eclipse.jetty,org.eclipse.jetty.tests,org.eclipse.jetty.websocket</includeGroupIds>
```
Comment (sbordet) [inline jetty-home/pom.xml]: Shouldn't the `async-rest` groupId be here too?

File: jetty-home/src/main/resources/modules/demo-base.mod [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+http
+https
+http2
+test-keystore
+work
+demo-root
+demo-async-rest
+demo-proxy
+demo-spec
+
```
Comment (sbordet) [inline jetty-home/src/main/resources/modules/demo-base.mod]: Unnecessary empty line.

File: jetty-start/src/main/java/org/eclipse/jetty/start/fileinits/MavenLocalRepoFileInitializer.java [hunk 9]
```diff
@@ -219,6 +290,65 @@ public Coordinates getCoordinates(URI uri)
         return coords;
     }
 
+    protected void download(Coordinates coords, Path destination)
+        throws IOException
+    {
+        if (coords.version.endsWith("-SNAPSHOT"))
```
Comment (sbordet) [inline jetty-start/src/main/java/org/eclipse/jetty/start/fileinits/MavenLocalRepoFileInitializer.java:296]: I don't think this is necessary at all - seems to introduce a lot of additional complexity.
Maybe it's a generic feature but I don't think it should go into this PR.

File: tests/test-webapps/test-jaas-webapp/pom.xml [hunk 10]
```diff
@@ -18,50 +18,37 @@
         <groupId>org.apache.maven.plugins</groupId>
         <artifactId>maven-deploy-plugin</artifactId>
         <configuration>
-          <!-- DO NOT DEPLOY (or Release) -->
-          <skip>true</skip>
+          <!-- DEMO Project, Must Deploy -->
+          <skip>false</skip>
         </configuration>
```
Comment (sbordet) [inline tests/test-webapps/test-jaas-webapp/pom.xml]: Just remove the `<configuration>` section altogether.

File: tests/test-webapps/test-jetty-webapp/pom.xml [hunk 11]
```diff
@@ -88,50 +86,55 @@
...
-        </configuration>
-      </plugin>
       <plugin>
         <groupId>org.apache.maven.plugins</groupId>
         <artifactId>maven-deploy-plugin</artifactId>
         <configuration>
-          <!-- must deploy: required for jetty-distribution -->
+          <!-- DEMO Project, Must Deploy -->
           <skip>false</skip>
         </configuration>
```
Comment (sbordet) [inline tests/test-webapps/test-jetty-webapp/pom.xml]: Remove `<configuration>` section.

File: tests/test-webapps/test-jetty-webapp/pom.xml [hunk 12]
```diff
@@ -88,50 +86,55 @@
...
+              <loginService implementation="org.eclipse.jetty.security.HashLoginService">
+                <name>Test Realm</name>
+                <config>src/main/config/demo-base/etc/realm.properties</config>
+              </loginService>
+            </loginServices>
+          </configuration>
+        </plugin>
+
+      </plugins>
+    </pluginManagement>
```
Comment (sbordet) [inline tests/test-webapps/test-jetty-webapp/pom.xml]: Why has this part been moved from `<build>` to `<pluginManagement>`? We don't need to manage plugins for children POMs, no?

File: tests/test-webapps/test-jetty-webapp/src/main/config/modules/demo.d/demo-jetty-override-web.xml [hunk 13]
```diff
@@ -0,0 +1,64 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<web-app 
+   xmlns="http://xmlns.jcp.org/xml/ns/javaee" 
+   xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+   xsi:schemaLocation="http://xmlns.jcp.org/xml/ns/javaee http://xmlns.jcp.org/xml/ns/javaee/web-app_3_1.xsd" 
+   version="3.1">
```
Comment (sbordet) [inline tests/test-webapps/test-jetty-webapp/src/main/config/modules/demo.d/demo-jetty-override-web.xml]: Use 4.0.

File: tests/test-webapps/test-jndi-webapp/pom.xml [hunk 14]
```diff
@@ -18,8 +18,8 @@
         <groupId>org.apache.maven.plugins</groupId>
         <artifactId>maven-deploy-plugin</artifactId>
         <configuration>
-          <!-- DO NOT DEPLOY (or Release) -->
-          <skip>true</skip>
+          <!-- DEMO Project, Must Deploy -->
+          <skip>false</skip>
         </configuration>
```
Comment (sbordet) [inline tests/test-webapps/test-jndi-webapp/pom.xml]: Remove `<configuration>` section.

File: tests/test-webapps/test-jndi-webapp/src/main/config/modules/demo.d/demo-jndi.xml [hunk 15]
```diff
@@ -0,0 +1,74 @@
+<?xml version="1.0"?><!DOCTYPE Configure PUBLIC "-//Jetty//Configure//EN" "https://www.eclipse.org/jetty/configure_9_3.dtd">
```
Comment (sbordet) [inline tests/test-webapps/test-jndi-webapp/src/main/config/modules/demo.d/demo-jndi.xml]: Must be `configure_10_0.dtd`.

File: tests/test-webapps/test-servlet-spec/test-spec-webapp/pom.xml [hunk 16]
```diff
@@ -18,7 +18,8 @@
         <groupId>org.apache.maven.plugins</groupId>
         <artifactId>maven-deploy-plugin</artifactId>
         <configuration>
-          <skip>true</skip>
+          <!-- DEMO Project, Must Deploy -->
+          <skip>false</skip>
         </configuration>
```
Comment (sbordet) [inline tests/test-webapps/test-servlet-spec/test-spec-webapp/pom.xml]: Remove `<configuration>`

**Review by sbordet — CHANGES_REQUESTED** [review]
Seems there are too many things in this PR - I thought it was just a couple of modules added?

File: examples/async-rest/async-rest-webapp/src/main/config/modules/demo-async-rest.mod (hunk 6 above)
Comment (joakime) [reply examples/async-rest/async-rest-webapp/src/main/config/modules/demo-async-rest.mod]: Done.

File: jetty-home/pom.xml (hunk 7 above)
Comment (joakime) [reply jetty-home/pom.xml]: Done.

File: jetty-home/src/main/resources/modules/demo-base.mod (hunk 8 above)
Comment (joakime) [reply jetty-home/src/main/resources/modules/demo-base.mod]: Done.

File: jetty-start/src/main/java/org/eclipse/jetty/start/fileinits/MavenLocalRepoFileInitializer.java (hunk 9 above)
Comment (joakime) [reply jetty-start/src/main/java/org/eclipse/jetty/start/fileinits/MavenLocalRepoFileInitializer.java:296]: As discussed, this is ok.

File: tests/test-webapps/test-jetty-webapp/pom.xml (hunk 12 above)
Comment (joakime) [reply tests/test-webapps/test-jetty-webapp/pom.xml]: Having jetty-maven-plugin as a formal plugin declaration forces it to resolve in the maven reactor.
It was causing a cyclic dependency as  ...

```
        WARNING: not enabled by default, as it introduces a cyclic reference in the build reactor.
        [ERROR] The projects in the reactor contain a cyclic reference:
          Edge between 'Vertex{label='org.eclipse.jetty:jetty-home:10.0.0-SNAPSHOT'}' and
          'Vertex{label='org.eclipse.jetty:test-jetty-webapp:10.0.0-SNAPSHOT'}'
          introduces to cycle in the graph:
           1: org.eclipse.jetty:test-jetty-webapp:10.0.0-SNAPSHOT
           2: org.eclipse.jetty:jetty-maven-plugin:10.0.0-SNAPSHOT
           3: org.eclipse.jetty:jetty-home:10.0.0-SNAPSHOT
           4: org.eclipse.jetty:test-jetty-webapp:10.0.0-SNAPSHOT @
```

Simply moving it to `<pluginManagement>` eliminates it from being involved in the reactor, but still allows it to be used from the command line at will.

File: tests/test-webapps/test-jetty-webapp/src/main/config/modules/demo.d/demo-jetty-override-web.xml (hunk 13 above)
Comment (joakime) [reply tests/test-webapps/test-jetty-webapp/src/main/config/modules/demo.d/demo-jetty-override-web.xml]: Oddly, no other web descriptor in `jetty-10.0.x` uses the correct XML intro.

From: https://javaee.github.io/servlet-spec/downloads/servlet-4.0/servlet-4_0_FINAL.pdf

``` xml
<web-app xmlns="http://xmlns.jcp.org/xml/ns/javaee"
xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
xsi:schemaLocation="http://xmlns.jcp.org/xml/ns/javaee webapp_4_0.xsd"
version="4.0">
```

Changing this breaks things.
Lets work this out in a different PR.

Comment (joakime) [reply tests/test-webapps/test-jetty-webapp/src/main/config/modules/demo.d/demo-jetty-override-web.xml]: Opened as Issue #5319

File: tests/test-webapps/test-jndi-webapp/src/main/config/modules/demo.d/demo-jndi.xml (hunk 15 above)
Comment (joakime) [reply tests/test-webapps/test-jndi-webapp/src/main/config/modules/demo.d/demo-jndi.xml]: Fixed in 1feb60bb2c8e521f80aaee2abc901c067b561026 (for many files in `/demos/`)

File: tests/test-webapps/test-servlet-spec/test-spec-webapp/pom.xml (hunk 16 above)
Comment (joakime) [reply tests/test-webapps/test-servlet-spec/test-spec-webapp/pom.xml]: Fixed

**Pull request merged**